submariner/multiple-cable-drivers.md

```diff
@@ -0,0 +1,132 @@
+cluster set. For example, a cluster set could contain three clusters, two that are on-premise and the other
+in the cloud. Connections to the cloud cluster will need to use encryption while the on-premise clusters can use
+unencrypted connections, as they’re connected by a private link, to avoid the performance overhead of encryption.
+This could be achieved using different cable driver implementations or the same cable driver implementation with
+differing encryption configurations.
+
+## Proposal
+
+This enhancement proposes a mechanism to allow users to specify their intent as to which clusters to connect and how to
+connect them. The default would still be a full mesh with an homogeneous cable driver.
+
+The standard mechanism in Kubernetes to specify intent is by creating a Kubernetes API resource whose effect is eventually
+realized into the cluster's desired state. To extend the Kubernetes API, one could either use the generic Kubernetes
+`ConfigMap` or use a custom resource definition (CRD). There's pros and cons to each but a CRD seems more suitable for
+this enhancement.
+
+This following CRD is proposed:
+
+```Go
+type ClusterConnectivityPolicies struct {
+    metav1.TypeMeta `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    Spec ClusterConnectivityPoliciesSpec `json:"spec"`
+}
+
+type ClusterConnectivityPoliciesSpec struct {
+    Policies []ClusterConnectivityPolicy `json:"policies,omitempty"`
+}
+
+type ClusterConnectivityPolicy struct {
+    // LeftClusterSelector identifies the Cluster resources representing the clusters on one end of a connection. An empty
+    // selector indicates wildcard, ie matches any cluster.
+    LeftClusterSelector metav1.LabelSelector `json:"leftClusterSelector,omitempty"`
+
+    // RightClusterSelector identifies the Cluster resources representing the clusters on the other end of a connection.
+    // An empty selector indicates wildcard, ie matches any cluster.
+    RightClusterSelector metav1.LabelSelector `json:"rightClusterSelector,omitempty"`
+
+    CableDriver *CableDriverSpec `json:"cableDriver,omitempty"`
+}
+
+type CableDriverSpec struct {
+    // Name of the cable driver implementation.
+    Name string `json:"name"`
+
+    // Options specifies key/value configuration parameters for the cable driver.
+    Options map[string]string `json:"options,omitempty"`
+}
+```
+
+The `ClusterConnectivityPolicies` CRD defines a list of policies that are applied in order of precedence. To resolve the
+connectivity policy for a remote cluster's `Endpoint`, Submariner would iterate the list of policies to search for a policy
+whose left and right label selectors match the corresponding `Cluster` resources of the local and remote `Endpoints`. The
+first matching policy is used. If no matching policy is found, the clusters are not connected.
+
+For full mesh connectivity, a policy with empty left and right label selectors would match all clusters. This would be the
```

Comment:
Should we be providing the option to decide topology per connection? Wouldn't be more realistic to have a uniform topology within a clusterset. Like a clusterset with full-mesh , another one with hub and spoke.

Reply:
A topology can be realized in this manner. The last example below is hub-and-spoke. One could also define multiple hub-and-spokes.

Reply:
That said, we could instead allow it to be defined in terms of specific topologies, eg

```
spec:
  policies:
     - hub-and-spoke
         hubSelector:
             database-role: server
         spokeSelector:
             database-role: client
      - full-mesh
```

I'm not sure what other topologies there could be. Also we want to be able to satisfy the use case for specifying the cable driver for a set of connections (as in the second example below).  The proposed API here makes it flexible.

Reply:
By clusterset I meant  a group of clusters as defined in [1]. So if we are planning to align with this, I think it would be easier to have uniform topology within a clusterset as it will be easier to program routes.

[1]https://github.com/kubernetes/enhancements/tree/master/keps/sig-multicluster/2149-clusterid